platforms/core-execution/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/AbstractFileWatcherUpdater.java

```diff
@@ -141,41 +143,28 @@ private void update(SnapshotHierarchy root) {
             updateWatchesOnChangedWatchedFiles(watchedFiles);
         }
 
-        // Probe every hierarchy that is watched, even ones nested inside others
-        ImmutableSet<File> oldProbedHierarchies = probedHierarchies;
-        probedHierarchies = watchableHierarchies.stream()
+        ImmutableSet<File> probedHierarchies = watchableHierarchies.stream()
             .filter(watchedFiles::contains)
             .collect(ImmutableSet.toImmutableSet());
-        if (oldProbedHierarchies.equals(probedHierarchies)) {
-            return;
-        }
 
-        oldProbedHierarchies.stream()
-            .filter(oldProbedHierarchy -> !probedHierarchies.contains(oldProbedHierarchy))
-            .forEach(probedHierarchy -> {
-                File probeDirectory = probeRegistry.getProbeDirectory(probedHierarchy);
-                probeRegistry.disarmWatchProbe(probedHierarchy);
-                stopWatchingProbeDirectory(probeDirectory);
-            });
-
-        probedHierarchies.stream()
-            .filter(newProbedHierarchy -> !oldProbedHierarchies.contains(newProbedHierarchy))
-            .forEach(probedHierarchy -> {
-                File probeDirectory = probeRegistry.getProbeDirectory(probedHierarchy);
-                // Make sure the directory exists, this can be necessary when
-                // included builds are evaluated with configuration cache
-                //noinspection ResultOfMethodCallIgnored
-                probeDirectory.mkdirs();
-                startWatchingProbeDirectory(probeDirectory);
-                probeRegistry.armWatchProbe(probedHierarchy);
-            });
+        probeRegistry.updateProbedHierarchies(
+            probedHierarchies,
+            this::stopWatchingProbeDirectory,
+            (probeLocation, isSubdirectoryOfWatchedHierarchy) -> {
+            // Make sure the directory exists, this can be necessary when
+            // included builds are evaluated with configuration cache
+            //noinspection ResultOfMethodCallIgnored
+                probeLocation.mkdirs();
```

Comment:
❓ To me `probeDirectory` means "the directory that will hold the probe," but `probeLocation` indicated "this is the path that we'll use as the probe." I think we should either keep using `probeDirectory` or use `probeLocation` to indicate the path to the actual probe. WDYT?